src/main/java/edu/ucla/library/iiif/fester/handlers/PostThumbnailsHandler.java

```diff
@@ -0,0 +1,207 @@
+    private static final String ID_TAG = "@id";
+
+    private final String myExceptionPage;
+
+    private final String myFesterizeVersion;
+
+    /**
+     * FUA = Festerize User Agent
+     */
+    private final Pattern myFUAPattern;
+
+    /**
+     * @param aVertx
+     * @param aConfig
+    */
+    public PostThumbnailsHandler(final Vertx aVertx, final JsonObject aConfig) throws IOException {
+        super(aVertx, aConfig);
+
+        final byte[] bytes = IOUtils.readBytes(getClass().getResourceAsStream("/webroot/error.html"));
+
+        myExceptionPage = new String(bytes, StandardCharsets.UTF_8);
+        myFesterizeVersion = aConfig.getString(Config.FESTERIZE_VERSION);
+        myFUAPattern = Pattern.compile("Festerize/(?<version>\\d+\\.\\d+\\.\\d+)");
+    }
+
+    /*
+     * (non-Javadoc)
+     * @see io.vertx.core.Handler#handle(java.lang.Object)
+     */
+    @Override
+    public void handle(final RoutingContext aContext) {
+        final HttpServerRequest request = aContext.request();
+        final HttpServerResponse response = aContext.response();
+        final Set<FileUpload> csvUploads = aContext.fileUploads();
+        final String festerizeUserAgent = StringUtils.trimTo(request.getHeader("User-Agent"), Constants.EMPTY);
+        final Matcher festerizeUserAgentMatcher = myFUAPattern.matcher(festerizeUserAgent);
+        final String errorMessage;
+
+        // An uploaded CSV is required
+        if (csvUploads.isEmpty()) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_037);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else if (festerizeUserAgentMatcher.matches() &&
+                   !festerizeUserAgentMatcher.group("version").equals(myFesterizeVersion)) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_147, myFesterizeVersion);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else {
+            final FileUpload csvFile = csvUploads.iterator().next();
+            final String filePath = csvFile.uploadedFileName();
+            final String fileName = csvFile.fileName();
+
+            try {
+                final List<String[]> originalLines =
+                    new CSVReader(Files.newBufferedReader(Paths.get(filePath))).readAll();
+                final List<String[]> linesWithThumbs = ThumbnailUtils.addThumbnailColumn(originalLines);
+                final int manifestIndex = Arrays.asList(linesWithThumbs.get(0)).indexOf(CSV.MANIFEST_URL);
+                final String collectionURL = linesWithThumbs.get(1)[manifestIndex];
+                final Future<JsonObject> collection = getManifest(collectionURL);
+                collection.onComplete(result -> {
+                    if (result.failed()) {
+                        returnError(response, HTTP.INTERNAL_SERVER_ERROR, result.cause().getMessage());
+                    } else {
+                        final JsonArray canvases = result.result().getJsonArray(MANIFESTS);
```

Comment:
This expression evaluates to an array of Manifests, not Canvases

(Even then it is specific to version 2.x of the IIIF Presentation API. Are we ready to support both API versions?)

Reply:
I think "yes" to "Are we ready to support both API versions."

Reply:
Some additional pointers that may be useful:

- To determine the API version, you can query the `@context` property of a Manifest or Collection
  - [v2](https://iiif.io/api/presentation/2.1/#linked-data-context-and-extensions)
  - [v3](https://iiif.io/api/presentation/3.0/#46-linked-data-context-and-extensions)
- For v2, the list of Canvases is at `manifest["sequences"][0]["canvases"]` (since all our v2 manifests should have only one Sequence)
- For v3, the list of Canvases is at `manifest["items"]`